### PR TITLE
feat: CartoCiudad (IGN España) geocoding provider

### PR DESCRIPTION
**Pull request by bgoniasa**
## What

Adds CartoCiudad as a geocoding provider — the Spanish national geocoder by IGN/CNIG.

## Why

GeoLibre already supports Nominatim, Pelias, ArcGIS, Mapbox, and Google. CartoCiudad is the best option for Spanish addresses: official data from Catastro/Correos/IGN, including `refCatastral`, `portalNumber`, `tip_via`, `postalCode`, `comunidadAutónoma`. No API key required.

## Changes

- `packages/core/src/geocoding.ts`:
  - Added `cartociudad` to `GeocodingProviderId` union type
  - New `cartociudadProvider` implementing the `GeocodingProvider` interface
  - Forward: `/geocoder/api/geocoder/find?q=`
  - Reverse: `/geocoder/api/geocoder/reverseGeocode?lat=&lon=`
  - `requiresApiKey: false`, `acceptsApiKey: false`
  - Rich address parsing: `tip_via`, `address`, `portalNumber`, `postalCode`, `poblacion`, `muni`, `province`, `comunidadAutonoma`, `refCatastral`

- `tests/geocoding.test.ts`: 9 tests covering metadata, URL building, forward/reverse parsing, edge cases (null input, missing coordinates)

## API format

CartoCiudad returns a single JSON object (not an array). Verified live against the API:

```json
{
  "lat": 40.416461,
  "lng": -3.704658,
  "address": "MAYOR",
  "tip_via": "CALLE",
  "portalNumber": 1,
  "postalCode": "28013",
  "poblacion": "Madrid",
  "refCatastral": "0343302VK4704C",
  "type": "portal",
  "state": 0
}
```

## Tests

All 4234 tests pass (including 9 new CartoCiudad tests).

## Notes

- CartoCiudad API returns a single JSON object (not an array)
- No confidence score available → `score: null`
- CORS allowed by IGN servers for browser requests
- Coverage limited to Spain (expected — it is a Spanish national service)

## Summary by CodeRabbit

* **New Features**
  * Added CartoCiudad (IGN España) as a supported geocoding provider.
  * Added forward and reverse geocoding with structured address details and readable location names.
  * Added validation for incomplete results, invalid coordinates, and API-key requirements.

* **Bug Fixes**
  * Improved handling of empty responses and no-content results.
  * Normalized longitudes outside the standard range before reverse-geocoding requests.

* **Documentation**
  * Updated configuration guidance with the CartoCiudad provider option.

### CODE REVIEW

File: packages/core/src/geocoding.ts [hunk 1]
```diff
@@ -661,13 +667,130 @@ function parseGoogleResults(data: unknown): GeocodeMatch[] {
...
+    return url.toString();
+  },
+  parseForward: (data) => {
+    // CartoCiudad `find` returns a single object or null, not an array.
+    if (!data || typeof data !== "object") return [];
+    const r = data as CartoCiudadResult;
+    const lat = Number(r.lat);
+    const lon = Number(r.lng);
+    if (!Number.isFinite(lat) || !Number.isFinite(lon)) return [];
+    return [
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/core/src/geocoding.ts]: Already fixed in 2af44ff0, which added the explicit `null`/`undefined` guard ahead of `Number()`. Hardened further in 170c7f34: both parsers now go through a single `cartociudadCoords` helper that rejects null/blank/non-numeric (`"NaN"`, `"invalid"`) values and anything outside WGS84 bounds.

On `state`: I checked the live API. A query it cannot match returns **HTTP 204 with an empty body**, not an object with a failure code, and `state` is `0` on every match observed (including ambiguous and out-of-range-portal queries). So `state` is not the success signal — I dropped it from the interface and documented the 204 contract in its place. The 204 was itself a real bug: `response.json()` on an empty body throws, so a nonsense query surfaced as "Search failed" instead of no match. 170c7f34 handles 204/empty bodies as no results in `geocodeForward`/`geocodeReverse`.

File: packages/core/src/geocoding.ts [hunk 2]
```diff
@@ -661,13 +667,130 @@ function parseGoogleResults(data: unknown): GeocodeMatch[] {
...
+  comunidadAutonoma?: string;
+  refCatastral?: string;
+  state?: number;
+  [key: string]: unknown;
+}
+
+/** Default endpoints for the public CartoCiudad REST API (no auth required). */
+const CARTOCIUDAD_FORWARD_ENDPOINT = "https://www.cartociudad.es/geocoder/api/geocoder/find";
+const CARTOCIUDAD_REVERSE_ENDPOINT =
+  "https://www.cartociudad.es/geocoder/api/geocoder/reverseGeocode";
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/core/src/geocoding.ts:740]: Thanks — taking the three in order:

1. **`geom` JSDoc**: stale, 2af44ff0 already rewrote that doc comment; there is no `geom` mention left on `CartoCiudadResult`.
2. **Module doc**: also stale — CartoCiudad was added to that list in 2af44ff0 (line 15 reads "ArcGIS World Geocoder, CartoCiudad, Mapbox, Pelias, Google").
3. **`limit`**: correct, it was a no-op. Fixed in 170c7f34 — `buildForwardUrl` no longer forwards it and carries a comment in the same style as the Google provider.

File: packages/core/src/geocoding.ts [hunk 3]
```diff
@@ -661,13 +667,130 @@ function parseGoogleResults(data: unknown): GeocodeMatch[] {
...
+    if (!Number.isFinite(lat) || !Number.isFinite(lon)) return [];
+    return [
+      {
+        lat,
+        lon,
+        displayName: cartociudadDisplayName(r),
+        score: null,
+      },
+    ];
+  },
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/core/src/geocoding.ts:808]: Good question to raise — I checked it against the live endpoints rather than guessing, and the "no match" path is different from what's hypothesized here:

- A query CartoCiudad cannot match returns **HTTP 204 with an empty body**, not a well-formed object with `type: "unknown"` or a bad `state`. Same for `reverseGeocode` on a point outside Spain.
- `state` is `0` on every match I could produce (exact portal, ambiguous street, nonexistent portal number, reverse in Madrid), and `stateMsg` is empty. It is not a success/failure signal worth filtering on.

That 204 turned out to be a genuine bug: `response.ok` is true for 204, so `await response.json()` threw on the empty body and a nonsense query surfaced as "Search failed" rather than "no match". 170c7f34 adds `readGeocodeJson`, which treats 204/empty bodies as no results for all providers, plus a test. I also dropped the unused `state` field and documented the 204 contract on the interface so the next reader doesn't go looking for an in-body flag.

The `geom` nit is stale — 2af44ff0 already removed that line from the JSDoc.

File: packages/core/src/geocoding.ts [hunk 4]
```diff
@@ -886,6 +1017,16 @@ export async function geocodeReverse(
...
+  if (
+    !Number.isFinite(lon) ||
+    !Number.isFinite(lat) ||
+    lon < -180 ||
+    lon > 180 ||
+    lat < -90 ||
+    lat > 90
+  ) {
+    return null;
+  }
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/core/src/geocoding.ts]: Agreed, and this is the best catch of the round — thank you. A click on a repeated world copy does arrive unwrapped, and rejecting it regressed reverse geocoding for every provider, not just CartoCiudad.

Fixed in 170c7f34: `geocodeReverse` now wraps longitude into [-180, 180] via a `wrapLongitude` helper (same formula as `LngLat.wrap()`) and passes the wrapped value to `buildReverseUrl`; only a non-finite longitude or an out-of-range latitude short-circuits to `null` now. Added a test asserting that `geocodeReverse(210, 48.85, …)` reaches the provider as `lon=-150` instead of being dropped.

File: packages/core/src/geocoding.ts [hunk 5]
```diff
@@ -661,13 +667,138 @@ function parseGoogleResults(data: unknown): GeocodeMatch[] {
...
+  parseForward: (data) => {
+    // CartoCiudad `find` returns a single object or null, not an array.
+    if (!data || typeof data !== "object") return [];
+    const r = data as CartoCiudadResult;
+    if (r.lat === null || r.lat === undefined || r.lng === null || r.lng === undefined) return [];
+    if (typeof r.lat === "string" && (r.lat as string).trim() === "") return [];
+    if (typeof r.lng === "string" && (r.lng as string).trim() === "") return [];
+    const lat = Number(r.lat);
+    const lon = Number(r.lng);
+    if (!Number.isFinite(lat) || !Number.isFinite(lon)) return [];
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/core/src/geocoding.ts]: Both checked against the live API:

1. **Arrays**: `find` returns a single object even for an ambiguous query — I tried `q=calle mayor`, which could plausibly return candidates, and it still answers with one object. Since an array is a shape we don't understand rather than a candidate list to pick the first of, 170c7f34 rejects it explicitly (`Array.isArray`) in both parsers instead of falling through, with a comment recording what was verified.
2. **`state`**: it is `0` on every match I could produce, and a genuine no-match is signalled by **HTTP 204 with an empty body**, not by a populated object with a bad `state`. So there is nothing to filter on. 170c7f34 drops the field from the interface and documents the 204 contract in its place — and handles that 204, which previously threw out of `response.json()`.

File: packages/core/src/geocoding.ts [hunk 6]
```diff
@@ -661,13 +667,138 @@ function parseGoogleResults(data: unknown): GeocodeMatch[] {
...
+  forward: true,
+  reverse: true,
+  requiresApiKey: false,
+  acceptsApiKey: false,
+  defaultForwardEndpoint: CARTOCIUDAD_FORWARD_ENDPOINT,
+  defaultReverseEndpoint: CARTOCIUDAD_REVERSE_ENDPOINT,
+  buildForwardUrl: (config, query, options) => {
+    const url = new URL(config.forwardEndpoint);
+    url.searchParams.set("q", query);
+    if (options.limit) url.searchParams.set("limit", String(options.limit));
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/core/src/geocoding.ts]: All three addressed in 170c7f34:

- **`limit`**: dropped from `buildForwardUrl`, with a comment in the same style as the Google provider (`find` has no result-count parameter).
- **Empty `portalNumber`**: now trimmed and skipped when blank, so `""` no longer appends a trailing space.
- **Duplication**: the lat/lng validation is now one `cartociudadCoords` helper called by both `parseForward` and `parseReverse`, alongside `cartociudadCoordinate` for the per-field coercion — same shape as `stringifyParts`/`coerceScore`.

File: packages/core/src/geocoding.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/core/src/geocoding.ts]: Valid on both counts — `parseReverse` in particular wasn't converting the coordinates at all, so `"NaN"` passed its blank check.

Fixed in 170c7f34: a shared `cartociudadCoords` helper now converts and validates for both parsers, rejecting non-numeric (`"NaN"`, `"invalid"`), non-finite, and out-of-WGS84-range values; `parseReverse` calls it before building the display result. Tests added in `tests/geocoding.test.ts` covering non-finite, non-numeric-string, and out-of-range latitude and longitude for both forward and reverse parsing.

File: packages/core/src/geocoding.ts [hunk 7]
```diff
@@ -852,6 +1007,36 @@ function geocodeFetch(): typeof globalThis.fetch {
...
+ * Normalize a longitude into [-180, 180], or null when it is not finite.
+ *
+ * MapLibre reports `lngLat.lng` unwrapped once the map is zoomed out far enough
+ * to show repeated world copies, so a click on a non-primary copy arrives as
+ * e.g. 210 rather than -150. Rejecting those outright would silently drop a
+ * legitimate reverse-geocode click for every provider, so wrap instead.
+ */
+function wrapLongitude(lon: number): number | null {
+  if (!Number.isFinite(lon)) return null;
+  return ((((lon + 180) % 360) + 360) % 360) - 180;
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/core/src/geocoding.ts:1096]: Confirmed and fixed in 73474373 — the drift is real for every ordinary longitude, not just an edge case:

```
-3.7     -> -3.7000000000000455
2.35     ->  2.3500000000000227
48.85    -> 48.85000000000002
-74.006  -> -74.00599999999997
```

Took the suggested early return. It also keeps `180` as `180` rather than flipping it to `-180`, which the unconditional form did. Extended the world-copy test to assert an in-range longitude reaches the provider unperturbed (`lon=-3.7`, not `-3.7000000000000455`).

File: packages/core/src/geocoding.ts [hunk 8]
```diff
@@ -661,13 +667,162 @@ function parseGoogleResults(data: unknown): GeocodeMatch[] {
...
+  const locality = [r.postalCode, r.poblacion]
+    .filter((s) => s && String(s).trim())
+    .map((s) => String(s).trim())
+    .join(" ");
+  if (locality) parts.push(locality);
+  if (r.province && String(r.province).trim()) parts.push(`(${r.province})`);
+  return parts.join(", ") || r.address || r.poblacion || "";
+}
+
+const cartociudadProvider: GeocodingProvider = {
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/core/src/geocoding.ts:774]: Deliberate, and I checked IGN's documentation rather than assuming. Leaving it as-is, with the reasoning now written down in `shouldThrottle`'s JSDoc (22859abf).

The 1 req/sec pace and 1000-row cap are **Nominatim's published usage policy**, not a general courtesy default — the module docstring already frames them that way. IGN documents no rate limit for the CartoCiudad REST service, no registration, and no key; the data is CC-BY 4.0. More telling: IGN's own bulk address geocoder is documented to process up to **60,000 records** per run (`unified_max_rows`), so batch usage at that scale is expected and supported by the service itself.

Applying Nominatim's numbers here would invent a constraint the service does not impose, and would stretch a 60k-row batch across roughly 18 hours. If IGN publishes a limit later, the fix is a real number from their policy rather than Nominatim's borrowed one.

Ref: https://github.com/IDEESpain/Cartociudad

File: packages/core/src/geocoding.ts [hunk 9]
```diff
@@ -661,13 +667,162 @@ function parseGoogleResults(data: unknown): GeocodeMatch[] {
...
+ * object (not an array) for both forward (`find`) and reverse (`reverseGeocode`)
+ * lookups, even for an ambiguous query. `lat`/`lng` are top-level numbers.
+ *
+ * There is no in-body success flag to check: a query the geocoder cannot match
+ * comes back as HTTP 204 with an empty body rather than an object carrying a
+ * failure code, so {@link readGeocodeJson} turns it into "no results" before a
+ * parser ever sees it. The body's `state` is 0 on every match observed against
+ * the live endpoints and is therefore not modeled here.
+ */
+interface CartoCiudadResult {
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/core/src/geocoding.ts:699]: Good instinct to push on an empirically-derived assumption — but the documented semantics turn out to be stronger than "always 0 so far", not weaker. From IGN's official service docs:

> `state`: 0 (este valor con la versión actual del geocoder, **se ha suprimido**, ya que se emplea elasticsearch y no se puede configurar la salida de candidates según grado de coincidencia).

Same wording for `stateMsg`. So `state` is not a confidence signal that happens to read 0 — it is a **vestigial field, removed** in the current Elasticsearch-backed geocoder precisely because that engine cannot rank candidates by degree of match. There is no municipality-centroid-vs-exact-address quality flag to lose.

The granularity signal that does exist is `type` (`portal`, `callejero`, `poblacion`), which is already surfaced in the reverse-geocode `parts`. `score: null` is correct for forward matches because the API exposes no ranking to map onto it.

22859abf replaces the empirical note in the JSDoc with the documented reason and the source link, so the assumption no longer rests on my probing.

Ref: https://github.com/IDEESpain/Cartociudad

File: packages/core/src/geocoding.ts [hunk 10]
```diff
@@ -661,13 +667,162 @@ function parseGoogleResults(data: unknown): GeocodeMatch[] {
...
+      r.portalNumber === undefined || r.portalNumber === null ? "" : String(r.portalNumber).trim();
+    if (portalNumber) street += ` ${portalNumber}`;
+    parts.push(street);
+  }
+  const locality = [r.postalCode, r.poblacion]
+    .filter((s) => s && String(s).trim())
+    .map((s) => String(s).trim())
+    .join(" ");
+  if (locality) parts.push(locality);
+  if (r.province && String(r.province).trim()) parts.push(`(${r.province})`);
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/core/src/geocoding.ts]: Correct, fixed in 22859abf. Pulled the trim into a local so the trimmed value is what gets interpolated, matching how the street and locality segments are built:

```ts
const province = r.province ? String(r.province).trim() : "";
if (province) parts.push(`(${province})`);
```

File: packages/core/src/geocoding.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/core/src/geocoding.ts:774]: Leaving this thread open rather than resolving it: the evidence says no throttle is warranted, but "how gently should we treat a free public government service" is a maintainer call, not a code defect. @giswqs — if you'd rather ship a conservative pace anyway, say so and I'll add CartoCiudad to `shouldThrottle` with its own constant rather than reusing Nominatim's.

File: packages/core/src/geocoding.ts [hunk 11]
```diff
@@ -739,6 +899,13 @@ export function resolveGeocoderConfig(input: GeocodingPreferenceInput): Geocoder
...
  * keyed provider) returns false.
+ *
+ * CartoCiudad is deliberately not included even though it is the other free,
+ * unkeyed public host: the 1 req/sec pace and 1000-row cap are Nominatim's
+ * published policy, not a general one. IGN documents no rate limit, and its own
+ * bulk address geocoder processes up to 60,000 records per run, so applying
+ * Nominatim's numbers would invent a constraint the service does not impose and
+ * stretch a 60k batch across 18 hours.
  */
 export function shouldThrottle(endpoint: string): boolean {
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/core/src/geocoding.ts]: You're right and I was wrong — the 60,000-record figure describes IGN's purpose-built bulk geocoder (`unified_max_rows`), not the single-lookup `find`/`reverseGeocode` endpoints this provider actually calls. Citing it as evidence that the interactive endpoints tolerate an unthrottled loop was a conflation.

Changed in 955b5f0a. CartoCiudad now gets its own conservative pace:

```ts
export const CARTOCIUDAD_MIN_INTERVAL_MS = 200; // ~5 req/sec, our courtesy default
```

Deliberately **not** Nominatim's 1100ms — that number is OSMF's published policy and borrowing it would still be inventing a constraint. 200ms is documented in the constant's JSDoc as ours, not IGN's, precisely because the server's burst behavior is unverified.

It stays **out of `shouldThrottle`**, so no 1000-row cap: IGN documents no row limit, and `shouldThrottle` is the Nominatim-policy gate (throttle *and* cap), not a general "is this paced" flag. To keep that distinction honest I split the hostname lookup into a `geocoderHostname` helper so pacing and the policy gate no longer share one boolean, and clarified both JSDocs. The user guide's usage-policy section documents the pace too.

Interactive place search is unaffected — `LayerPanelPlaceSearch` already takes `Math.max(DEBOUNCE_MS = 500, interval)`.

File: packages/core/src/geocoding.ts [hunk 12]
```diff
@@ -661,13 +667,167 @@ function parseGoogleResults(data: unknown): GeocodeMatch[] {
...
+    let street = streetParts.join(" ");
+    const portalNumber =
+      r.portalNumber === undefined || r.portalNumber === null ? "" : String(r.portalNumber).trim();
+    if (portalNumber) street += ` ${portalNumber}`;
+    parts.push(street);
+  }
+  const locality = [r.postalCode, r.poblacion]
+    .filter((s) => s && String(s).trim())
+    .map((s) => String(s).trim())
+    .join(" ");
```
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/core/src/geocoding.ts:767]: Good catch, and it reproduces. Probing the live API for a result with `muni` but no `poblacion`, a road kilometre point does exactly that:

```
q=A-6 km 120 -> {type: portal, address: 'A-6', postalCode: '05296',
                 poblacion: None, muni: 'Espinosa de los Caballeros', province: 'Ávila'}
```

So the display name rendered as "A-6, 05296, (Ávila)" — the municipality was present in the response and silently dropped.

Fixed in 955b5f0a: the locality segment now falls back to `muni` when `poblacion` is blank, and the final bare-name fallback gained `muni` as well. `poblacion` still wins when both are present, since it is the more specific locality ("Cal Mayoral" within municipality "l'Espunyola"). Both cases have tests built from the live response shapes.

File: packages/core/src/geocoding.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (giswqs) [reply packages/core/src/geocoding.ts:774]: Updating my earlier reply: I've reversed the decision and implemented a pace in 955b5f0a. A follow-up comment correctly pointed out that the 60,000-record figure I cited describes IGN's purpose-built bulk geocoder, not the single-lookup `find`/`reverseGeocode` endpoints this provider calls — so "unbounded is fine" was an unverified assumption, not a documented one.

CartoCiudad now paces at `CARTOCIUDAD_MIN_INTERVAL_MS = 200` (~5 req/sec), documented as our own courtesy default rather than IGN policy. It keeps no row cap and does not borrow Nominatim's 1100ms. Resolving this thread; the reasoning lives in the JSDoc and the user guide.
